src/sage/algebras/weyl_algebra.py

```diff
@@ -1123,3 +1141,519 @@ def _act_(self, g, x):
         D = {y: c for (y, dy), c in f.monomial_coefficients(copy=False).items()
              if all(dyi == 0 for dyi in dy)}
         return self.right_domain()(D)
+
+
+class InfGenDifferentialWeylAlgebraElement(IndexedFreeModuleElement):
+    """
+    An element of an infinitely generated differential Weyl algebra.
+
+    TESTS::
+
+        sage: W.<x> = DifferentialWeylAlgebra(QQ, n=oo)
+        sage: W.inject_variables(verbose=False)
+        sage: W.zero() == 0
+        True
+        sage: W.one() == 1
+        True
+        sage: W.zero() == 1
+        False
+        sage: x[1] == 1
+        False
+        sage: W(x[1]) == x[1]
+        True
+        sage: W(dx[1]) == dx[1]
+        True
+        sage: W(x[1] + dx[2]^2) == x[1] + dx[2]^2
+        True
+        sage: x[1] == x[11]
+        False
+        sage: x[1] / 2
+        1/2*x[1]
+        sage: W(2) / 2
+        1
+        sage: (x[1] + dx[1]) * (-4/3)
+        -4/3*dx[1] - 4/3*x[1]
+        sage: (-4/3) * (x[1] + dx[1])
+        -4/3*dx[1] - 4/3*x[1]
+    """
+    def _repr_(self) -> str:
+        """
+        Return a string representation of ``self``.
+
+        TESTS::
+
+            sage: W.<x> = DifferentialWeylAlgebra(QQ, n=oo)
+            sage: dx = W.differentials()
+            sage: dx[1]^2*x[1]^2
+            x[1]^2*dx[1]^2 + 4*x[1]*dx[1] + 2
+        """
+        def term(m):
+            res = ''
+            if not m[0].is_one():
+                res += m[0]._repr_()
+            if not m[1].is_one():
+                if res != '':
+                    res += '*'
+                res += m[1]._repr_()
+            return res if res != '' else '1'
+        return repr_from_monomials(self.list(), term)
+
+    def _mul_(self, other):
+        """
+        Multiply ``self`` by ``other``.
+
+        TESTS::
+
+            sage: W.<x> = DifferentialWeylAlgebra(QQ, n=oo)
+            sage: dx = W.differentials()
+            sage: dx[1]*(x[1]*x[2] + x[3])
+            x[1]*x[2]*dx[1] + x[3]*dx[1] + x[2]
+            sage: dx[132]*x[132] - x[132]*dx[132] == 1
+            True
+        """
+        out = {}
+        zero = self.parent().base_ring().zero()
+        zero_m = self.parent()._diff_index.one()
+        # multiply each pair of monomials
+        for ml in self._monomial_coefficients:
+            cl = self._monomial_coefficients[ml]
+            for mr in other._monomial_coefficients:
+                cr = other._monomial_coefficients[mr]
+                # apply derivative terms of ml to mr and simplify
+                cur = [((mr[0], zero_m), cl*cr)]
+                ldd = ml[1].dict()
+
+                for i in ldd:
+                    p = ldd[i]
+                    # applying dx[i] p times
+                    for _ in range(p):
+                        next = []
+                        for m, c in cur:
+                            diff = dict(m[1].dict())
```

Comment:
```suggestion
                            diff = m[1].dict()
```
The extra `dict` call should not be necessary.